multicluster/controllers/multicluster/node_controller.go

```diff
@@ -0,0 +1,173 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	mcsv1alpha1 "antrea.io/antrea/multicluster/apis/multicluster/v1alpha1"
+	"antrea.io/antrea/multicluster/controllers/multicluster/common"
+)
+
+type (
+	// NodeReconciler is for member cluster only.
+	// It will create a Gateway object if a Node has an annotation `multicluster.antrea.io/gateway:true`
+	// and update corresponding Gateway if any subnets changes.
+	NodeReconciler struct {
+		client.Client
+		Scheme     *runtime.Scheme
+		namespace  string
+		precedence mcsv1alpha1.Precedence
+	}
+)
+
+// NewNodeReconciler creates a  NodeReconciler to watch Node object changes and create a
+// corresponding Gateway if the Node has the annotation `multicluster.antrea.io/gateway:true`.
+func NewNodeReconciler(
+	Client client.Client,
+	Scheme *runtime.Scheme,
+	namespace string,
+	precedence mcsv1alpha1.Precedence) *NodeReconciler {
+	if string(precedence) == "" {
+		precedence = mcsv1alpha1.PrecedencePrivate
+	}
+	reconciler := &NodeReconciler{
+		Client:     Client,
+		Scheme:     Scheme,
+		namespace:  namespace,
+		precedence: precedence,
+	}
+	return reconciler
+}
+
+//+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=gateways,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups="",resources=nodes,verbs=get;list;watch;
+//+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=gateways/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=gateways/finalizers,verbs=update
+
+func (r *NodeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	klog.V(2).InfoS("Reconciling Node", "node", req.Name)
+	gw := &mcsv1alpha1.Gateway{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      req.Name,
+			Namespace: r.namespace,
+		},
+	}
+	// When the Node is annotated with 'multicluster.antrea.io/gateway=true' as a Gateway:
+	//   - Delete the Gateway if the Node is deleted
+	//   - Update the Gateway if Node's InternalIP or GatewayIP is updated
+	//   - Create a new Gateway if there is no existing Gateway
+	node := &corev1.Node{}
+	if err := r.Client.Get(ctx, req.NamespacedName, node); err != nil {
+		if apierrors.IsNotFound(err) {
+			err := r.Client.Delete(ctx, gw, &client.DeleteOptions{})
+			return ctrl.Result{}, client.IgnoreNotFound(err)
+		}
+		klog.ErrorS(err, "Failed to get Node", "node", req.Name)
+		return ctrl.Result{}, err
+	}
+
+	_, isGW := node.Annotations[common.GatewayAnnotation]
+	var err error
+	var gwIP, internalIP string
+	if internalIP, gwIP, err = r.getGatawayNodeIP(node); err != nil {
+		klog.ErrorS(err, "There is no valid Gateway IP for Node, will retry later when there is any new Node update", "node", node.Name)
+		return ctrl.Result{}, nil
+	}
+	gwNamespacedName := types.NamespacedName{
+		Name:      node.Name,
+		Namespace: r.namespace,
+	}
+
+	if err := r.Client.Get(ctx, gwNamespacedName, gw); err != nil {
```

Comment:
I saw you chose to remove the cache, which means we will have an API call to get Gateway for each Node change (Status change, etc.). It might not be too bad, but I still wonder what is the standard implementation with kube-builder.

What you think we use Builder.Owns() to to watch Gateways and let Gateway changes trigger Reconcile as well?

Reply:
This `get` will actually hit local shared cache, it's not a real API call, this is wrapped by kube-builder framework. https://github.com/kubernetes-sigs/controller-runtime/blob/3f265c36d7bf597facc9b0363b48ff21a68cb782/doc.go#L78-L92. when you see `r.Client` it's actually not the Client provide by `client-go`.

Reply:
I believe that should be the case for watched objects, but non-watched objects I doubt the client knows what to cache. And how can it guarantee the cache has the latest state?

Reply:
According to the doc, when the request is triggered first time, it will also set up corresponding cache `The Cache will auto-populate with watched objects, as well as when other structured objects are requested.`, I think the default sync time is 10 hours.  https://github.com/kubernetes-sigs/controller-runtime/blob/3f265c36d7bf597facc9b0363b48ff21a68cb782/pkg/cache/cache.go#L145

Reply:
The question is if Gateway changes, will the cache be auto-refreshed, when we do not watch Gateway. Could you check that?

And if that really happens, a further question is how can the client make sure you can get the latest Gateway (e.g. maybe some changes are not received yet) to compare with.

Reply:
According to the codes here https://github.com/kubernetes-sigs/controller-runtime/blob/3f265c36d7bf597facc9b0363b48ff21a68cb782/pkg/cache/internal/informers_map.go#L190-L214, I think it will set up cache for un-watched objects when it's requested, and it will check `HasSynced()` to make sure it doesn't read from a stale cache.

Reply:
Thanks for the info. But I still wonder how the cache can guarantee Get can return the latest state. @tnqn : what are your thoughts?

Reply:
sure, I will check the framework codes and get beck to you.

Reply:
When controller intialize resource reconciler, it will use `mgr.GetClient()`, the sample code is here: https://github.com/antrea-io/antrea/blob/b01ba8ea8f7311ca8e76331cc9e11fd80a73a168/multicluster/cmd/multicluster-controller/member.go#L64
By default, manager's client is actually calling `Cluster` interface's [`GetClient()` ](https://github.com/kubernetes-sigs/controller-runtime/blob/3f265c36d7bf597facc9b0363b48ff21a68cb782/pkg/cluster/cluster.go#L51-L55), The `Cluster`'s default `NewClient` will create a [`DelegatingClient`](https://github.com/kubernetes-sigs/controller-runtime/blob/3f265c36d7bf597facc9b0363b48ff21a68cb782/pkg/cluster/cluster.go#L258-L270) by default. When we call like `r.Client.Get()`, it's actually calling `DelegatingClient`'s `Get()` [here](https://github.com/kubernetes-sigs/controller-runtime/blob/3f265c36d7bf597facc9b0363b48ff21a68cb782/pkg/client/split.go#L123-L131), as you can see, it's eventually hit the cacheReader to read cache to get objects. The default new cache creation function is [here](https://github.com/kubernetes-sigs/controller-runtime/blob/3f265c36d7bf597facc9b0363b48ff21a68cb782/pkg/cache/cache.go#L147-L169). The cache will create `informerCache`, and cache's `Get()` will hit `informerCache`'s `Get()` [here](https://github.com/kubernetes-sigs/controller-runtime/blob/3f265c36d7bf597facc9b0363b48ff21a68cb782/pkg/cache/informer_cache.go#L53-L69). It will create a new [SharedIndexInformer](https://github.com/kubernetes-sigs/controller-runtime/blob/3f265c36d7bf597facc9b0363b48ff21a68cb782/pkg/cache/internal/informers_map.go#L216-L265) with default 10 hours resync period if the object informer is not found.  I think we can update default 10hours to a shorter time to make sure it's synced in time.

Reply:
My question is - how can we guarantee what we read from the cache is the latest state. For example:

1.  At t1, Reconciler created Gateway1 for Node1.
2. At t2, Reconciler received Node1 delete, but the client cache is not updated with the newly created Gateway1, and so Reconciler would not do anything to delete Gateway1.
3. At t3, the client cache received Gateway1, but no event will trigger deletion of Gateway1 any more.

Reply:
Only Get/List are operations based on cache, there also writer interface, Delete action will actually go to API Server directly. when Reconciler get a Node delete event, it will always call Delete() to API Server to delete corresponding GW and ignore not found error.

Reply:
My point is your code compares the desired state with the state returned by Get() to decide if a update/delete is required. I hope you to evaluate if that is safe. Maybe for delete, you do not compare with Get() (but then you will always need an API call even it is not necessary), but please check all the cases.

Reply:
yes, you are right, I noticed that there is a comparison to use `Get()` check if create is necessary or not, it may raise Gateway already exists error. but when there is an error, the event will be requeued and handle in next round, eventually the cache will be updated, the event will do a Gateway update if necessary. According to the doc, it indeed will be slightly out of date: `In general, write your controller with the assumption that information will eventually be correct, but may be slightly out of date. ` https://github.com/kubernetes-sigs/controller-runtime/blob/3f265c36d7bf597facc9b0363b48ff21a68cb782/FAQ.md

Reply:
I have same question here. `HasSynced` can only ensure the cache has been synced at least once, and doesn't mean the cache is in sync with server. I think the problem @jianjuns raised in https://github.com/antrea-io/antrea/pull/3463/files#r874350144 is valiad unless `Client.Create()` also updates cache on API call success. But I don't find the method will update cache.

The `Owns` method looks like the solution for this case:
> Owns defines types of Objects being *generated* by the ControllerManagedBy, and configures the ControllerManagedBy to respond to create / delete / update events by *reconciling the owner object*

Reply:
Just check through the code again. The problem is more for Gateway update (when IPs change). One way is in this PR, we do not check the existing Gateway and always update, and we optimize in a follow-up PR.

Reply:
@tnqn , thanks for the advice, I will investigate`Owns` method and underlying cache in controller-runtime, but I refined the code without comparing IPs, so it will always update the Gateway as @jianjuns  suggested for now, thanks